File: docx_parsers/styles/paragraph_properties_parser.py
# paragraph_properties_parser.py

import xml.etree.ElementTree as ET
from typing import Optional
from docx_parsers.helpers.common_helpers import extract_element, extract_attribute
from docx_parsers.utils import convert_twips_to_points
from docx_parsers.models.styles_models import ParagraphStyleProperties, SpacingProperties, IndentationProperties

class ParagraphPropertiesParser:
    """
    A parser for extracting paragraph properties from an XML element.
    """

    def parse(self, pPr_element: ET.Element) -> ParagraphStyleProperties:
        """
        Parses paragraph properties from the given XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            ParagraphStyleProperties: The parsed paragraph style properties.
        """
        properties = ParagraphStyleProperties()
        
        if pPr_element is not None:
            properties.spacing = self.extract_spacing(pPr_element)
            properties.indent = self.extract_indentation(pPr_element)
            properties.outline_level = self.extract_outline_level(pPr_element)
            properties.widow_control = self.extract_widow_control(pPr_element)
            properties.suppress_auto_hyphens = self.extract_suppress_auto_hyphens(pPr_element)
            properties.bidi = self.extract_bidi(pPr_element)
            properties.justification = self.extract_justification(pPr_element)
            properties.keep_next = self.extract_keep_next(pPr_element)
            properties.suppress_line_numbers = self.extract_suppress_line_numbers(pPr_element)

        return properties

    def extract_spacing(self, pPr_element: ET.Element) -> Optional[SpacingProperties]:
        """
        Extracts spacing properties from the paragraph properties XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            Optional[SpacingProperties]: The extracted spacing properties or None.
        """
        spacing_element = extract_element(pPr_element, "w:spacing")
        if spacing_element is not None:
            spacing_properties = SpacingProperties()
            before = extract_attribute(spacing_element, 'before')
            after = extract_attribute(spacing_element, 'after')
            line = extract_attribute(spacing_element, 'line')
            if before:
                spacing_properties.before_pt = convert_twips_to_points(int(before))
            if after:
                spacing_properties.after_pt = convert_twips_to_points(int(after))
            if line:
                spacing_properties.line_pt = convert_twips_to_points(int(line))
            return spacing_properties
        return None

    def extract_indentation(self, pPr_element: ET.Element) -> Optional[IndentationProperties]:
        """
        Extracts indentation properties from the paragraph properties XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            Optional[IndentationProperties]: The extracted indentation properties or None.
        """
        indent_element = extract_element(pPr_element, "w:ind")
        if indent_element is not None:
            indent_properties = IndentationProperties()
            left = extract_attribute(indent_element, 'left') or extract_attribute(indent_element, 'start')
            right = extract_attribute(indent_element, 'right') or extract_attribute(indent_element, 'end')
            hanging = extract_attribute(indent_element, 'hanging')
            firstLine = extract_attribute(indent_element, 'firstLine')
            if left:
                indent_properties.left_pt = convert_twips_to_points(int(left))
            if right:
                indent_properties.right_pt = convert_twips_to_points(int(right))
            if hanging:
                indent_properties.hanging_pt = convert_twips_to_points(int(hanging))
            if firstLine:
                indent_properties.firstline_pt = convert_twips_to_points(int(firstLine))
            return indent_properties
        return None

    def extract_outline_level(self, pPr_element: ET.Element) -> Optional[int]:
        """
        Extracts the outline level from the paragraph properties XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            Optional[int]: The extracted outline level or None.
        """
        outline_lvl_element = extract_element(pPr_element, "w:outlineLvl")
        if outline_lvl_element is not None:
            outline_level = extract_attribute(outline_lvl_element, 'val')
            if outline_level is not None:
                return int(outline_level)
        return None

    def extract_widow_control(self, pPr_element: ET.Element) -> Optional[bool]:
        """
        Extracts the widow control property from the paragraph properties XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            Optional[bool]: The extracted widow control property or None.
        """
        widow_control_element = extract_element(pPr_element, "w:widowControl")
        if widow_control_element is not None:
            return True
        return None

    def extract_suppress_auto_hyphens(self, pPr_element: ET.Element) -> Optional[bool]:
        """
        Extracts the suppress auto hyphens property from the paragraph properties XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            Optional[bool]: The extracted suppress auto hyphens property or None.
        """
        suppress_auto_hyphens_element = extract_element(pPr_element, "w:suppressAutoHyphens")
        if suppress_auto_hyphens_element is not None:
            return True
        return None

    def extract_bidi(self, pPr_element: ET.Element) -> Optional[bool]:
        """
        Extracts the bidi property from the paragraph properties XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            Optional[bool]: The extracted bidi property or None.
        """
        bidi_element = extract_element(pPr_element, "w:bidi")
        if bidi_element is not None:
            return True
        return None

    def extract_justification(self, pPr_element: ET.Element) -> Optional[str]:
        """
        Extracts the justification property from the paragraph properties XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            Optional[str]: The extracted justification property or None.
        """
        justification_element = extract_element(pPr_element, "w:jc")
        if justification_element is not None:
            return extract_attribute(justification_element, 'val')
        return None

    def extract_keep_next(self, pPr_element: ET.Element) -> Optional[bool]:
        """
        Extracts the keep next property from the paragraph properties XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            Optional[bool]: The extracted keep next property or None.
        """
        keep_next_element = extract_element(pPr_element, "w:keepNext")
        if keep_next_element is not None:
            return True
        return None

    def extract_suppress_line_numbers(self, pPr_element: ET.Element) -> Optional[bool]:
        """
        Extracts the suppress line numbers property from the paragraph properties XML element.

        Args:
            pPr_element (ET.Element): The paragraph properties XML element.

        Returns:
            Optional[bool]: The extracted suppress line numbers property or None.
        """
        suppress_line_numbers_element = extract_element(pPr_element, "w:suppressLineNumbers")
        if suppress_line_numbers_element is not None:
            return True
        return None

File: docx_parser_converter_ts/src/docx_parsers/styles/paragraphPropertiesParser.ts
import type {
  ParagraphStyleProperties,
  SpacingProperties,
  IndentationProperties,
} from '../models/stylesModels';
import {
  extractElement,
  extractAttribute,
  extractBooleanAttribute,
  safeInt,
} from '../helpers/commonHelpers';
import { convertTwipsToPoints } from '../utils';

/**
 * Parses the paragraph properties from a DOCX paragraph properties element (w:pPr).
 *
 * This class extracts and parses various properties related to paragraph formatting,
 * converting them into structured TypeScript interfaces for further processing or
 * conversion to other formats.
 */
export class ParagraphPropertiesParser {
  /**
   * Parses the given paragraph properties element into a ParagraphStyleProperties object.
   *
   * @param pPrElement - The paragraph properties XML element (w:pPr).
   * @returns The parsed paragraph style properties.
   *
   * @example
   * The following is an example of a paragraph properties element:
   * ```xml
   * <w:pPr>
   *   <w:spacing w:before="240" w:after="240" w:line="360"/>
   *   <w:ind w:left="720" w:right="720" w:firstLine="720"/>
   *   <w:jc w:val="both"/>
   *   <w:outlineLvl w:val="1"/>
   *   <w:widowControl/>
   *   <w:suppressAutoHyphens/>
   *   <w:bidi/>
   *   <w:keepNext/>
   *   <w:suppressLineNumbers/>
   * </w:pPr>
   * ```
   */
  public parse(pPrElement: Element | null): ParagraphStyleProperties {
    const properties: ParagraphStyleProperties = {};

    if (pPrElement) {
      properties.spacing = this.extractSpacing(pPrElement);
      properties.indent = this.extractIndentation(pPrElement);
      properties.outlineLevel = this.extractOutlineLevel(pPrElement);
      properties.widowControl = this.extractWidowControl(pPrElement);
      properties.suppressAutoHyphens = this.extractSuppressAutoHyphens(pPrElement);
      properties.bidi = this.extractBidi(pPrElement);
      properties.justification = this.extractJustification(pPrElement);
      properties.keepNext = this.extractKeepNext(pPrElement);
      properties.suppressLineNumbers = this.extractSuppressLineNumbers(pPrElement);
    }

    return properties;
  }

  /**
   * Extracts spacing properties from the given paragraph properties element.
   *
   * @param pPrElement - The paragraph properties element.
   * @returns The extracted spacing properties, or undefined if not present.
   * ```xml
   * <w:spacing w:before="240" w:after="240" w:line="360"/>
   * ```
   */
  private extractSpacing(pPrElement: Element): SpacingProperties | undefined {
    const spacingElement = extractElement(pPrElement, ".//w:spacing");
    if (spacingElement) {
      const spacingProperties: SpacingProperties = {};
      const before = extractAttribute(spacingElement, 'before');
      const after = extractAttribute(spacingElement, 'after');
      const line = extractAttribute(spacingElement, 'line');

      const beforeVal = safeInt(before);
      if (beforeVal !== null) {
        spacingProperties.beforePt = convertTwipsToPoints(beforeVal);
      }
      const afterVal = safeInt(after);
      if (afterVal !== null) {
        spacingProperties.afterPt = convertTwipsToPoints(afterVal);
      }
      const lineVal = safeInt(line);
      if (lineVal !== null) {
        spacingProperties.linePt = convertTwipsToPoints(lineVal);
      }
      return Object.keys(spacingProperties).length > 0 ? spacingProperties : undefined;
    }
    return undefined;
  }

  /**
   * Extracts indentation properties from the given paragraph properties element.
   *
   * @param pPrElement - The paragraph properties element.
   * @returns The extracted indentation properties, or undefined if not present.
   * ```xml
   * <w:ind w:left="720" w:right="720" w:firstLine="720"/>
   * ```
   */
  private extractIndentation(pPrElement: Element): IndentationProperties | undefined {
    const indentElement = extractElement(pPrElement, ".//w:ind");
    if (indentElement) {
      let leftPt = this.convertToPoints(indentElement, ['left', 'start']);
      let rightPt = this.convertToPoints(indentElement, ['right', 'end']);
      let hangingPt = this.convertToPoints(indentElement, ['hanging']);
      let firstLinePt = this.convertToPoints(indentElement, ['firstLine']);

      if (hangingPt !== null && hangingPt !== undefined) {
        firstLinePt = -hangingPt;
      }
      
      const indentationProperties: IndentationProperties = {};
      if (leftPt !== null && leftPt !== undefined) indentationProperties.leftPt = leftPt;
      if (rightPt !== null && rightPt !== undefined) indentationProperties.rightPt = rightPt;
      if (firstLinePt !== null && firstLinePt !== undefined) indentationProperties.firstLinePt = firstLinePt;

      return Object.keys(indentationProperties).length > 0 ? indentationProperties : undefined;
    }
    return undefined;
  }

  /**
   * Converts attribute values found in an element to points.
   * It checks a list of attribute names and uses the first one found.
   *
   * @param element - The XML element containing the attributes.
   * @param attrs - The list of attribute names to check (e.g., ['left', 'start']).
   * @returns The converted value in points, or null if no attribute is found or if conversion fails.
   */
  private convertToPoints(element: Element, attrs: string[]): number | null {
    for (const attr of attrs) {
      const valueStr = extractAttribute(element, attr);
      if (valueStr !== null) {
        const valueInt = safeInt(valueStr);
        if (valueInt !== null) {
          return convertTwipsToPoints(valueInt);
        }
      }
    }
    return null;
  }

  /**
   * Extracts the outline level from the given paragraph properties element.
   *
   * @param pPrElement - The paragraph properties element.
   * @returns The extracted outline level, or undefined if not present.
   * ```xml
   * <w:outlineLvl w:val="1"/>
   * ```
   */
  private extractOutlineLevel(pPrElement: Element): number | undefined {
    const outlineLvlElement = extractElement(pPrElement, ".//w:outlineLvl");
    if (outlineLvlElement) {
      const outlineLevelStr = extractAttribute(outlineLvlElement, 'val');
      const outlineLevelInt = safeInt(outlineLevelStr);
      if (outlineLevelInt !== null) {
        return outlineLevelInt;
      }
    }
    return undefined;
  }

  /**
   * Extracts the widow control setting from the given paragraph properties element.
   *
   * @param pPrElement - The paragraph properties element.
   * @returns The widow control setting (boolean), or undefined if not specified.
   * ```xml
   * <w:widowControl/>
   * <!-- or -->
   * <w:widowControl w:val="true"/>
   * ```
   */
  private extractWidowControl(pPrElement: Element): boolean | undefined {
    const widowControlElement = extractElement(pPrElement, ".//w:widowControl");
    const val = extractBooleanAttribute(widowControlElement);
    return val === null ? undefined : val; // Convert null to undefined for consistency
  }

  /**
   * Extracts the suppress auto hyphens setting.
   *
   * @param pPrElement - The paragraph properties element.
   * @returns The suppress auto hyphens setting, or undefined.
   * ```xml
   * <w:suppressAutoHyphens/>
   * ```
   */
  private extractSuppressAutoHyphens(pPrElement: Element): boolean | undefined {
    const suppressElement = extractElement(pPrElement, ".//w:suppressAutoHyphens");
    const val = extractBooleanAttribute(suppressElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts the bidirectional setting.
   *
   * @param pPrElement - The paragraph properties element.
   * @returns The bidirectional setting, or undefined.
   * ```xml
   * <w:bidi/>
   * ```
   */
  private extractBidi(pPrElement: Element): boolean | undefined {
    const bidiElement = extractElement(pPrElement, ".//w:bidi");
    const val = extractBooleanAttribute(bidiElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts and maps the justification value.
   *
   * @param pPrElement - The paragraph properties element.
   * @returns The mapped justification value ('left', 'right', 'center', 'justify'), or undefined.
   * ```xml
   * <w:jc w:val="both"/>
   * ```
   */
  private extractJustification(pPrElement: Element): string | undefined {
    const jcElement = extractElement(pPrElement, ".//w:jc");
    if (jcElement) {
      const justificationVal = extractAttribute(jcElement, 'val');
      if (justificationVal !== null) {
        const mapping: { [key: string]: string } = {
          left: 'left',
          start: 'left',
          right: 'right',
          end: 'right',
          center: 'center',
          both: 'justify',
        };
        return mapping[justificationVal] || 'left'; // Default to left if unknown
      }
    }
    return undefined;
  }

  /**
   * Extracts the keep next setting.
   *
   * @param pPrElement - The paragraph properties element.
   * @returns The keep next setting, or undefined.
   * ```xml
   * <w:keepNext/>
   * ```
   */
  private extractKeepNext(pPrElement: Element): boolean | undefined {
    const keepNextElement = extractElement(pPrElement, ".//w:keepNext");
    const val = extractBooleanAttribute(keepNextElement);
    return val === null ? undefined : val;
  }

  /**
   * Extracts the suppress line numbers setting.
   *
   * @param pPrElement - The paragraph properties element.
   * @returns The suppress line numbers setting, or undefined.
   * ```xml
   * <w:suppressLineNumbers/>
   * ```
   */
  private extractSuppressLineNumbers(pPrElement: Element): boolean | undefined {
    const suppressElement = extractElement(pPrElement, ".//w:suppressLineNumbers");
    const val = extractBooleanAttribute(suppressElement);
    return val === null ? undefined : val;
  }
}

